ts2phc.8
.TH TS2PHC 8 "December 2019" "linuxptp"
.SH NAME
ts2phc - Synchronizes one or more PTP Hardware Clocks using external time stamps.

.SH SYNOPSIS
.B ts2phc
[
.B \-hmqv
] [
.BI \-c " device|name"
] [
.BI \-f " config"
] [
.BI \-l " print-level"
] [
.BI \-s " device|name"
] [
.I long-options
]
.I .\|.\|.

.SH DESCRIPTION
.B ts2phc
synchronizes PTP Hardware Clocks (PHC) to external time stamp signals.
A single source may be used to distribute time to one or more PHC devices.

.SH OPTIONS
.TP
.BI \-a
Adjust the direction of synchronization automatically. The program determines
which PHC should be a source of time and which should be a sink by querying the
port states from the running instance of
.B ptp4l.
Note that using this option, the PPS signal distribution hierarchy still
remains fixed as per the configuration file. This implies that using this
option, a PHC PPS master may become a time sink, and a PPS slave may become a
time source. Other, non-PHC types of PPS masters (generic, NMEA) cannot become
time sinks. Clocks which are not part of
.B ptp4l's
list of ports are not synchronized. This option is useful when the
.B boundary_clock_jbod
option of ptp4l is also enabled.
.TP
.BI \-c " device|name"
Specifies a PHC slave clock to be synchronized.
The clock may be identified by its character device (like /dev/ptp0)
or its associated network interface (like eth0).
This option may be given multiple times.
.TP
.BI \-f " config"
Read configuration from the specified file.
No configuration file is read by default.
.TP
.BI \-h
Displays the command line help summary.
.TP
.BI \-l " print-level"
Sets the maximum syslog level of messages which should be printed or
sent to the system logger. The default is 6 (LOG_INFO).
.TP
.B \-m
Prints log messages to the standard output.
.TP
.B \-q
Prevents sending log messages to the system logger.
.TP
.BI \-s " device|name"
Specifies the source of the PPS signal.
Use the key word "generic" for an external 1-PPS without ToD information.
When using a master PHC, the clock may be identified by its character
device (like /dev/ptp0) or its associated network interface (like
eth0).
Use the key word "nmea" for an external 1-PPS from a GPS providing ToD
information via the RMC NMEA sentence.
.TP
.B \-v
Prints the software version and exits.

.SH LONG OPTIONS

Each and every configuration file option (see below) may also appear
as a "long" style command line argument.  For example, the use_syslog
option may be set using either of these two forms.

.RS
\f(CW\-\-use_syslog 1   \-\-use_syslog=1\fP
.RE

Option values given on the command line override values in the global
section of the configuration file.

.SH CONFIGURATION FILE

The configuration file is divided into sections. Each section starts with a
line containing its name enclosed in brackets and it follows with settings.
Each setting is placed on a separate line, it contains the name of the
option and the value separated by whitespace characters. Empty lines and lines
starting with # are ignored.

There are two different section types.

.TP
.B 1.
The global section (indicated as
.BR [global] )
sets the program options and default slave clock options. Other
sections are clock specific sections, and they override the default
options.
.TP
.B 2.
Slave clock sections give the name of the configured slave (e.g.
.BR [eth0] ).
Slave clocks specified in the configuration file need not be specified
with the
.B \-c
command line option.

.SH GLOBAL OPTIONS

.TP
.B first_step_threshold
The maximum offset, specified in seconds, that the servo will correct
by changing the clock frequency instead of stepping the clock. This is
only applied on the first update. When set to 0.0, the servo will not
step the clock on start.
The default is 0.00002 (20 microseconds).
.TP
.B free_running
When set to 1, no the slave clock will be adjusted.
This option can be useful in test scenarios, for example to determine
how well synchronized a group of local clocks are to each other.
The default is 0 (adjust the slave clocks).
.TP
.B logging_level
The maximum logging level of messages which should be printed.
The default is 6 (LOG_INFO).
.TP
.B max_frequency
The maximum allowed frequency adjustment of the clock in parts per
billion.  This is an additional limit to the maximum allowed by the
hardware. When set to 0, the hardware limit will be used.
The default is 900000000 (90%).
.TP
.B message_tag
The tag which is added to all messages printed to the standard output
or system log.  The default is an empty string (which cannot be set in
the configuration file as the option requires an argument).
.TP
.B step_threshold
The maximum offset, specified in seconds, that the servo will correct
by changing the clock frequency instead of stepping the clock. When
set to 0.0, the servo will never step the clock except on start.
The default is 0.0.
.TP
.B ts2phc.nmea_remote_host, ts2phc.nmea_remote_port
Specifies the serial port character device providing ToD information
when using the "nmea" PPS signal source.  Note that if these two
options are both specified, then the given remote connection will be
used in preference to the configured serial port.
These options default to the empty string, that is, not specified.
.TP
.B ts2phc.nmea_serialport
Specifies the serial port character device providing ToD information
when using the "nmea" PPS signal source.  Note that if the options,
ts2phc.nmea_remote_host and ts2phc.nmea_remote_port, are both
specified, then the given remote connection will be used in preference
to the configured serial port.
The default is "/dev/ttyS0".
.TP
.B ts2phc.pulsewidth
The expected pulse width of the external PPS signal in nanoseconds.
When 'ts2phc.extts_polarity' is "both", the given pulse width is used
to detect and discard the time stamp of the unwanted edge.
The supported range is 1000000 to 990000000 nanoseconds.
The default is 500000000 nanoseconds.
.TP
.B use_syslog
Print messages to the system log if enabled.  The default is 1 (enabled).
.TP
.B verbose
Print messages to the standard output if enabled.  The default is 0 (disabled).

.SH SLAVE CLOCK OPTIONS

.TP
.B ts2phc.channel
The external time stamping or periodic output channel to be used.
Some PHC devices feature programmable pins and one or more time
stamping channels.  This option allows selecting a particular channel
to be used.  When using a PHC device as the PPS source, this option
selects the periodic output channel.
The default is channel 0.
.TP
.B ts2phc.extts_correction
The value, in nanoseconds, to be added to each PPS time stamp.
The default is 0 (no correction).
.TP
.B ts2phc.extts_polarity
The polarity of the external PPS signal, either "rising" or "falling".
Some PHC devices always time stamp both edges.  Setting this option to
"both" will allow the ts2phc program to work with such devices by
detecting and ignoring the unwanted edge.  In this case be sure to
set 'ts2phc.pulsewidth' to the correct value.
The default is "rising".
.TP
.B ts2phc.master
Setting this option to 1 configures the given PHC device as the source
of the PPS signal.
The default is 0 for the slave role.
.TP
.B ts2phc.pin_index
The pin index to be used.
Some PHC devices feature programmable pins, and this option allows
configuration of a particular pin for the external time stamping or
periodic output function.
The default is pin index 0.

.SH WARNING

Be cautious when sharing the same configuration file between ptp4l,
phc2sys, and ts2phc.  Keep in mind that values specified in the
configuration file take precedence over the default values.  If an
option which is common to the other programs is set in the
configuration file, then the value will be applied to all the programs
using the file, and this might not be what is expected.

It is recommended to use separate configuration files for ptp4l,
phc2sys, and ts2phc in order to avoid any unexpected behavior.

.SH SEE ALSO
.BR phc2sys (8)
.BR ptp4l (8)


ts2phc.c
/**
 * @file ts2phc.c
 * @brief Utility program to synchronize the PHC clock to external events
 * @note Copyright (C) 2013 Balint Ferencz <fernya@sch.bme.hu>
 * @note Based on the phc2sys utility
 * @note Copyright (C) 2012 Richard Cochran <richardcochran@gmail.com>
 * @note SPDX-License-Identifier: GPL-2.0+
 */
#include <stdlib.h>
#include <net/if.h>
#include <sys/types.h>
#include <unistd.h>

#include "clockadj.h"
#include "config.h"
#include "contain.h"
#include "interface.h"
#include "phc.h"
#include "print.h"
#include "ts2phc.h"
#include "version.h"

#define NS_PER_SEC		1000000000LL
#define SAMPLE_WEIGHT		1.0

struct interface {
	STAILQ_ENTRY(interface) list;
};

static void ts2phc_cleanup(struct ts2phc_private *priv)
{
	struct port *p, *tmp;

	ts2phc_slave_cleanup(priv);
	if (priv->master)
		ts2phc_master_destroy(priv->master);
	if (priv->cfg)
		config_destroy(priv->cfg);

	close_pmc_node(&priv->node);

	/*
	 * Clocks are destroyed by the cleanup methods of the individual
	 * master and slave PHC modules.
	 */
	LIST_FOREACH_SAFE(p, &priv->ports, list, tmp)
		free(p);

	msg_cleanup();
}

/* FIXME: Copied from phc2sys */
static int normalize_state(int state)
{
	if (state != PS_MASTER && state != PS_SLAVE &&
	    state != PS_PRE_MASTER && state != PS_UNCALIBRATED) {
		/* treat any other state as "not a master nor a slave" */
		state = PS_DISABLED;
	}
	return state;
}

/* FIXME: Copied from phc2sys */
static struct port *port_get(struct ts2phc_private *priv, unsigned int number)
{
	struct port *p;

	LIST_FOREACH(p, &priv->ports, list) {
		if (p->number == number)
			return p;
	}
	return NULL;
}

/* FIXME: Copied from phc2sys */
static int clock_compute_state(struct ts2phc_private *priv,
			       struct clock *clock)
{
	int state = PS_DISABLED;
	struct port *p;

	LIST_FOREACH(p, &priv->ports, list) {
		if (p->clock != clock)
			continue;
		/* PS_SLAVE takes the highest precedence, PS_UNCALIBRATED
		 * after that, PS_MASTER is third, PS_PRE_MASTER fourth and
		 * all of that overrides PS_DISABLED, which corresponds
		 * nicely with the numerical values */
		if (p->state > state)
			state = p->state;
	}
	return state;
}

#define node_to_ts2phc(node) \
	container_of(node, struct ts2phc_private, node)

static int ts2phc_recv_subscribed(struct pmc_node *node,
				  struct ptp_message *msg, int excluded)
{
	struct ts2phc_private *priv = node_to_ts2phc(node);
	int mgt_id, state;
	struct portDS *pds;
	struct port *port;
	struct clock *clock;

	mgt_id = get_mgt_id(msg);
	if (mgt_id == excluded)
		return 0;
	switch (mgt_id) {
	case TLV_PORT_DATA_SET:
		pds = get_mgt_data(msg);
		port = port_get(priv, pds->portIdentity.portNumber);
		if (!port) {
			pr_info("received data for unknown port %s",
				pid2str(&pds->portIdentity));
			return 1;
		}
		state = normalize_state(pds->portState);
		if (port->state != state) {
			pr_info("port %s changed state",
				pid2str(&pds->portIdentity));
			port->state = state;
			clock = port->clock;
			state = clock_compute_state(priv, clock);
			if (clock->state != state || clock->new_state) {
				clock->new_state = state;
				priv->state_changed = 1;
			}
		}
		return 1;
	}
	return 0;
}

struct servo *servo_add(struct ts2phc_private *priv, struct clock *clock)
{
	enum servo_type type = config_get_int(priv->cfg, NULL, "clock_servo");
	struct servo *servo;
	int fadj, max_adj;

	fadj = (int) clockadj_get_freq(clock->clkid);
	/* Due to a bug in older kernels, the reading may silently fail
	   and return 0. Set the frequency back to make sure fadj is
	   the actual frequency of the clock. */
	clockadj_set_freq(clock->clkid, fadj);

	max_adj = phc_max_adj(clock->clkid);

	servo = servo_create(priv->cfg, type, -fadj, max_adj, 0);
	if (!servo)
		return NULL;

	servo_sync_interval(servo, SERVO_SYNC_INTERVAL);

	return servo;
}

void clock_add_tstamp(struct clock *clock, tmv_t t)
{
	struct timespec ts = tmv_to_timespec(t);

	pr_debug("adding tstamp %ld.%09ld to clock %s",
		 ts.tv_sec, ts.tv_nsec, clock->name);
	clock->last_ts = t;
	clock->is_ts_available = 1;
}

static int clock_get_tstamp(struct clock *clock, tmv_t *ts)
{
	if (!clock->is_ts_available)
		return 0;
	clock->is_ts_available = 0;
	*ts = clock->last_ts;
	return 1;
}

static void clock_flush_tstamp(struct clock *clock)
{
	clock->is_ts_available = 0;
}

struct clock *clock_add(struct ts2phc_private *priv, const char *device)
{
	clockid_t clkid = CLOCK_INVALID;
	int phc_index = -1;
	struct clock *c;
	int err;

	clkid = posix_clock_open(device, &phc_index);
	if (clkid == CLOCK_INVALID)
		return NULL;

	LIST_FOREACH(c, &priv->clocks, list) {
		if (c->phc_index == phc_index) {
			/* Already have the clock, don't add it again */
			posix_clock_close(clkid);
			return c;
		}
	}

	c = calloc(1, sizeof(*c));
	if (!c) {
		pr_err("failed to allocate memory for a clock");
		return NULL;
	}
	c->clkid = clkid;
	c->phc_index = phc_index;
	c->servo_state = SERVO_UNLOCKED;
	c->servo = servo_add(priv, c);
	c->no_adj = config_get_int(priv->cfg, NULL, "free_running");
	err = asprintf(&c->name, "/dev/ptp%d", phc_index);
	if (err < 0) {
		free(c);
		posix_clock_close(clkid);
		return NULL;
	}

	LIST_INSERT_HEAD(&priv->clocks, c, list);
	return c;
}


void clock_destroy(struct clock *c)
{
	servo_destroy(c->servo);
	posix_clock_close(c->clkid);
	free(c->name);
	free(c);
}

/* FIXME: Copied from phc2sys */
static struct port *port_add(struct ts2phc_private *priv, unsigned int number,
			     char *device)
{
	struct clock *c = NULL;
	struct port *p, *tmp;

	p = port_get(priv, number);
	if (p)
		return p;
	/* port is a new one, look whether we have the device already on
	 * a different port */
	LIST_FOREACH(tmp, &priv->ports, list) {
		if (tmp->number == number) {
			c = tmp->clock;
			break;
		}
	}
	if (!c) {
		c = clock_add(priv, device);
		if (!c)
			return NULL;
	}
	p = malloc(sizeof(*p));
	if (!p) {
		pr_err("failed to allocate memory for a port");
		clock_destroy(c);
		return NULL;
	}
	p->number = number;
	p->clock = c;
	LIST_INSERT_HEAD(&priv->ports, p, list);
	return p;
}

static int auto_init_ports(struct ts2phc_private *priv)
{
	int state, timestamping;
	int number_ports, res;
	char iface[IFNAMSIZ];
	struct clock *clock;
	struct port *port;
	unsigned int i;

	while (1) {
		if (!is_running())
			return -1;
		res = run_pmc_clock_identity(&priv->node, 1000);
		if (res < 0)
			return -1;
		if (res > 0)
			break;
		/* res == 0, timeout */
		pr_notice("Waiting for ptp4l...");
	}

	number_ports = run_pmc_get_number_ports(&priv->node, 1000);
	if (number_ports <= 0) {
		pr_err("failed to get number of ports");
		return -1;
	}

	res = run_pmc_subscribe(&priv->node, 1000);
	if (res <= 0) {
		pr_err("failed to subscribe");
		return -1;
	}

	for (i = 1; i <= number_ports; i++) {
		res = run_pmc_port_properties(&priv->node, 1000, i, &state,
					      &timestamping, iface);
		if (res == -1) {
			/* port does not exist, ignore the port */
			continue;
		}
		if (res <= 0) {
			pr_err("failed to get port properties");
			return -1;
		}
		if (timestamping == TS_SOFTWARE) {
			/* ignore ports with software time stamping */
			continue;
		}
		port = port_add(priv, i, iface);
		if (!port)
			return -1;
		port->state = normalize_state(state);
	}
	if (LIST_EMPTY(&priv->clocks)) {
		pr_err("no suitable ports available");
		return -1;
	}
	LIST_FOREACH(clock, &priv->clocks, list) {
		clock->new_state = clock_compute_state(priv, clock);
	}
	priv->state_changed = 1;

	return 0;
}

static void ts2phc_synchronize_clocks(struct ts2phc_private *priv)
{
	struct timespec source_ts;
	tmv_t source_tmv;
	struct clock *c;
	int valid, err;

	err = ts2phc_master_getppstime(priv->master, &source_ts);
	if (err < 0) {
		pr_err("source ts not valid");
		return;
	}
	if (source_ts.tv_nsec > NS_PER_SEC / 2)
		source_ts.tv_sec++;
	source_ts.tv_nsec = 0;

	source_tmv = timespec_to_tmv(source_ts);

	LIST_FOREACH(c, &priv->clocks, list) {
		int64_t offset;
		double adj;
		tmv_t ts;

		valid = clock_get_tstamp(c, &ts);
		if (!valid) {
			pr_debug("%s timestamp not valid, skipping", c->name);
			continue;
		}

		offset = tmv_to_nanoseconds(tmv_sub(ts, source_tmv));

		if (c->no_adj) {
			pr_info("%s offset %10" PRId64, c->name,
				offset);
			continue;
		}

		adj = servo_sample(c->servo, offset, tmv_to_nanoseconds(ts),
				   SAMPLE_WEIGHT, &c->servo_state);

		pr_info("%s offset %10" PRId64 " s%d freq %+7.0f",
			c->name, offset, c->servo_state, adj);

		switch (c->servo_state) {
		case SERVO_UNLOCKED:
			break;
		case SERVO_JUMP:
			clockadj_set_freq(c->clkid, -adj);
			clockadj_step(c->clkid, -offset);
			break;
		case SERVO_LOCKED:
		case SERVO_LOCKED_STABLE:
			clockadj_set_freq(c->clkid, -adj);
			break;
		}
	}
}

static void usage(char *progname)
{
	fprintf(stderr,
		"\n"
		"usage: %s [options]\n\n"
		" -a             turn on autoconfiguration\n"
		" -c [dev|name]  phc slave clock (like /dev/ptp0 or eth0)\n"
		"                (may be specified multiple times)\n"
		" -f [file]      read configuration from 'file'\n"
		" -h             prints this message and exits\n"
		" -l [num]       set the logging level to 'num'\n"
		" -m             print messages to stdout\n"
		" -q             do not print messages to the syslog\n"
		" -s [dev|name]  source of the PPS signal\n"
		"                may take any of the following forms:\n"
		"                    generic   - an external 1-PPS without ToD information\n"
		"                    /dev/ptp0 - a local PTP Hardware Clock (PHC)\n"
		"                    eth0      - a local PTP Hardware Clock (PHC)\n"
		"                    nmea      - a gps device connected by serial port or network\n"
		" -v             prints the software version and exits\n"
		"\n",
		progname);
}

int main(int argc, char *argv[])
{
	int c, err = 0, have_slave = 0, index, print_level;
	char uds_local[MAX_IFNAME_SIZE + 1];
	enum ts2phc_master_type pps_type;
	struct ts2phc_private priv = {0};
	char *config = NULL, *progname;
	const char *pps_source = NULL;
	struct config *cfg = NULL;
	struct interface *iface;
	struct option *opts;
	int autocfg = 0;

	handle_term_signals();

	cfg = config_create();
	if (!cfg) {
		ts2phc_cleanup(&priv);
		return -1;
	}

	opts = config_long_options(cfg);

	/* Process the command line arguments. */
	progname = strrchr(argv[0], '/');
	progname = progname ? 1 + progname : argv[0];
	while (EOF != (c = getopt_long(argc, argv, "ac:f:hi:l:mqs:v", opts, &index))) {
		switch (c) {
		case 0:
			if (config_parse_option(cfg, opts[index].name, optarg)) {
				ts2phc_cleanup(&priv);
				return -1;
			}
			break;
		case 'a':
			autocfg = 1;
			break;
		case 'c':
			if (!config_create_interface(optarg, cfg)) {
				fprintf(stderr, "failed to add slave\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			have_slave = 1;
			break;
		case 'f':
			config = optarg;
			break;
		case 'l':
			if (get_arg_val_i(c, optarg, &print_level,
					  PRINT_LEVEL_MIN, PRINT_LEVEL_MAX)) {
				ts2phc_cleanup(&priv);
				return -1;
			}
			config_set_int(cfg, "logging_level", print_level);
			print_set_level(print_level);
			break;
		case 'm':
			config_set_int(cfg, "verbose", 1);
			print_set_verbose(1);
			break;
		case 'q':
			config_set_int(cfg, "use_syslog", 0);
			print_set_syslog(0);
			break;
		case 's':
			if (pps_source) {
				fprintf(stderr, "too many PPS sources\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			pps_source = optarg;
			break;
		case 'v':
			ts2phc_cleanup(&priv);
			version_show(stdout);
			return 0;
		case 'h':
			ts2phc_cleanup(&priv);
			usage(progname);
			return -1;
		case '?':
		default:
			ts2phc_cleanup(&priv);
			usage(progname);
			return -1;
		}
	}
	if (config && (c = config_read(config, cfg))) {
		fprintf(stderr, "failed to read config\n");
		ts2phc_cleanup(&priv);
		return -1;
	}
	print_set_progname(progname);
	print_set_tag(config_get_string(cfg, NULL, "message_tag"));
	print_set_verbose(config_get_int(cfg, NULL, "verbose"));
	print_set_syslog(config_get_int(cfg, NULL, "use_syslog"));
	print_set_level(config_get_int(cfg, NULL, "logging_level"));

	STAILQ_INIT(&priv.slaves);
	priv.cfg = cfg;

	snprintf(uds_local, sizeof(uds_local), "/var/run/ts2phc.%d",
		 getpid());

	if (autocfg) {
		err = init_pmc_node(cfg, &priv.node, uds_local,
				    ts2phc_recv_subscribed);
		if (err) {
			ts2phc_cleanup(&priv);
			return -1;
		}
		err = auto_init_ports(&priv);
		if (err) {
			ts2phc_cleanup(&priv);
			return -1;
		}
	}

	STAILQ_FOREACH(iface, &cfg->interfaces, list) {
		if (1 == config_get_int(cfg, interface_name(iface), "ts2phc.master")) {
			if (pps_source) {
				fprintf(stderr, "too many PPS sources\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			pps_source = interface_name(iface);
		} else {
			if (ts2phc_slave_add(&priv, interface_name(iface))) {
				fprintf(stderr, "failed to add slave\n");
				ts2phc_cleanup(&priv);
				return -1;
			}
			have_slave = 1;
		}
	}
	if (!have_slave) {
		fprintf(stderr, "no slave clocks specified\n");
		ts2phc_cleanup(&priv);
		usage(progname);
		return -1;
	}
	if (!pps_source) {
		fprintf(stderr, "no PPS source specified\n");
		ts2phc_cleanup(&priv);
		usage(progname);
		return -1;
	}
	if (ts2phc_slaves_init(&priv)) {
		fprintf(stderr, "failed to initialize slaves\n");
		ts2phc_cleanup(&priv);
		return -1;
	}

	if (!strcasecmp(pps_source, "generic")) {
		pps_type = TS2PHC_MASTER_GENERIC;
	} else if (!strcasecmp(pps_source, "nmea")) {
		pps_type = TS2PHC_MASTER_NMEA;
	} else {
		pps_type = TS2PHC_MASTER_PHC;
	}
	priv.master = ts2phc_master_create(&priv, pps_source, pps_type);
	if (!priv.master) {
		fprintf(stderr, "failed to create master\n");
		ts2phc_cleanup(&priv);
		return -1;
	}

	while (is_running()) {
		struct clock *c;

		LIST_FOREACH(c, &priv.clocks, list)
			clock_flush_tstamp(c);

		err = ts2phc_slave_poll(&priv);
		if (err < 0) {
			pr_err("poll failed");
			break;
		}
		if (err > 0)
			ts2phc_synchronize_clocks(&priv);
	}

	ts2phc_cleanup(&priv);
	return err;
}
